inspirehep/modules/workflows/tasks/refextract.py

```diff
@@ -86,49 +86,43 @@ def extract_journal_info(obj, eng):
     """Extract references from PDF and return in INSPIRE format."""
+
+    def field_builder(builder, value_l):
+        for value in value_l:
+            builder(value)
+
     references = extract_references_from_file(
         filepath,
         reference_format=u"{title},{volume},{page}",
     )
     mapped_references = []
-    if references.get('references'):
-        for ref in references.get('references'):
-            reference = {
-                'curated_relation': False,
-                'raw_refs': [
-                    {
-                        'position': '',
-                        'schema': '',
-                        'source': '',
-                        'value': json.dumps(ref),
-                    }
-                ],
-                'record': {
-                    '$ref': ''
-                },
-                'reference': {
-                    'authors': ref.get('author'),
-                    'book_series': ref.get('book_series'),
-                    'collaboration': ref.get('collaboration'),
-                    'dois': ref.get('dois'),
-                    'imprint': ref.get('imprint'),
-                    'misc': ref.get('misc'),
-                    'number': ref.get('linemarker'),
-                    'persistent_identifiers': ref.get(
-                        'persistent_identifiers'
-                    ),
-                    'publication_info': {
-                        'year': ref.get('year'),
-                        'journal_title': ref.get('journal_title'),
-                        'journal_volume': ref.get('journal_volume'),
-                        'page_start': ref.get('journal_page'),
-                        'artid': ref.get('journal_reference'),
-                    },
-                    'texkey': ref.get('texkey'),
-                    'titles': ref.get('titles'),
-                    'urls': ref.get('urls'),
-                },
-            }
-            mapped_references.append(reference)
+
+    if references:
+        for ref in references:
+            rb = ReferenceBuilder()
+            mapping = [
+                ('raw_ref', partial(field_builder, rb.add_raw_reference)),
+                ('author', partial(field_builder, rb.add_refextract_authors_str)),
+                ('collaboration', partial(field_builder, rb.add_collaboration)),
+                ('doi', partial(field_builder, rb.add_uid)),
+                ('texkey', partial(field_builder, rb.set_texkey)),
+                ('reportnumber', partial(field_builder, rb.add_report_number)),
+                ('title', partial(field_builder, rb.add_title)),
+                ('misc', partial(field_builder, rb.add_misc)),
+                ('publisher', partial(field_builder, rb.set_publisher)),
+                ('journal_reference', partial(field_builder, rb.set_pubnote)),
+                ('journal_year', partial(field_builder, rb.set_year)),
+                ('year', partial(field_builder, rb.set_year)),
```

Comment:
another option is to amend the builder: if the reference already contains a `publication_info/journal_title`, use `publication_info/year`, otherwise `imprint/date`.